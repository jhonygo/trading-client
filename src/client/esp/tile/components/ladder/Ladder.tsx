import React from 'react';
import style from './style.css';
import { Price } from '../../model/priceTick';

export namespace Ladder {
  export interface Props {
    bids: Price[];
    asks: Price[];
  }
}

const extractPips = price => price.substr(4, 2).padEnd(2, '00');

const Ladder: React.FC<Ladder.Props> = ({ bids = [], asks = [] }) => {
  const sellPrices = bids.map((p, i) => (
    <tr key={i}>
      <td className="w-25">
        <span className={style.quantity}>{p.quantity / 1000000}M</span>
      </td>
      <td className="w-75 text-right">
        <span className={style.bigFigures}>
          {p.price.toString().substr(0, 4)}
        </span>
        <span className={style.pips}>{extractPips(p.price.toString())}</span>
        <span className={style.tenthOfPips}>
          {p.price.toString().substr(6) || '0'}
        </span>
      </td>
    </tr>
  ));

  const buyPrices = asks.map((p, i) => (
    <tr key={i}>
      <td className="w-75">
        <span className={style.bigFigures}>
          {p.price.toString().substr(0, 4)}
        </span>
        <span className={style.pips}>{extractPips(p.price.toString())}</span>
        <span className={style.tenthOfPips}>
          {p.price.toString().substr(6) || '0'}
        </span>
      </td>
      <td className="text-right w-25">
        <span className={style.quantity}>{p.quantity / 1000000}M</span>
      </td>
    </tr>
  ));

  return (
    <div className="row">
      <div className="col-sm-6">
        <table className="table table-striped table-hover table-borderless no-footer">
          <tbody>{sellPrices}</tbody>
        </table>
      </div>
      <div className="col-sm-6">
        <table className="table table-striped table-hover table-borderless no-footer">
          <tbody>{buyPrices}</tbody>
        </table>
      </div>
    </div>
  );
};

export default Ladder;
