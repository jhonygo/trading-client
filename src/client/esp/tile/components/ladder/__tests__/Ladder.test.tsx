import React from 'react';
import { shallow } from 'enzyme';
import Ladder from '../Ladder';

describe('Ladder', () => {
  it('should render liquidity ladder', () => {
    // When
    const price = shallow(
      <Ladder
        asks={[
          { quantity: 1000000, price: 1.23127 },
          { quantity: 5000000, price: 1.26578 }
        ]}
        bids={[
          { quantity: 1000000, price: 1.23118 },
          { quantity: 5000000, price: 1.26571 }
        ]}
      />
    );

    // Then
    expect(price).toMatchSnapshot();
  });
});
