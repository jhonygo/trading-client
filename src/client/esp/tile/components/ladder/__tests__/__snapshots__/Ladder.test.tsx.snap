// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ladder should render liquidity ladder 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-6"
  >
    <table
      className="table table-striped table-hover table-borderless no-footer"
    >
      <tbody>
        <tr
          key="0"
        >
          <td
            className="w-25"
          >
            <span
              className="quantity"
            >
              1
              M
            </span>
          </td>
          <td
            className="w-75 text-right"
          >
            <span
              className="bigFigures"
            >
              1.23
            </span>
            <span
              className="pips"
            >
              11
            </span>
            <span
              className="tenthOfPips"
            >
              8
            </span>
          </td>
        </tr>
        <tr
          key="1"
        >
          <td
            className="w-25"
          >
            <span
              className="quantity"
            >
              5
              M
            </span>
          </td>
          <td
            className="w-75 text-right"
          >
            <span
              className="bigFigures"
            >
              1.26
            </span>
            <span
              className="pips"
            >
              57
            </span>
            <span
              className="tenthOfPips"
            >
              1
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-sm-6"
  >
    <table
      className="table table-striped table-hover table-borderless no-footer"
    >
      <tbody>
        <tr
          key="0"
        >
          <td
            className="w-75"
          >
            <span
              className="bigFigures"
            >
              1.23
            </span>
            <span
              className="pips"
            >
              12
            </span>
            <span
              className="tenthOfPips"
            >
              7
            </span>
          </td>
          <td
            className="text-right w-25"
          >
            <span
              className="quantity"
            >
              1
              M
            </span>
          </td>
        </tr>
        <tr
          key="1"
        >
          <td
            className="w-75"
          >
            <span
              className="bigFigures"
            >
              1.26
            </span>
            <span
              className="pips"
            >
              57
            </span>
            <span
              className="tenthOfPips"
            >
              8
            </span>
          </td>
          <td
            className="text-right w-25"
          >
            <span
              className="quantity"
            >
              5
              M
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
